src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,609 @@
+LXD_GROUP = "lxd"
+UBUNTU_HOME = Path("/home/ubuntu")
+
+
+def _configure_system_users() -> None:
+    """Configure system users.
+
+    Raises:
+        SystemUserConfigurationError: If there was an error configuring ubuntu user.
+    """
+    try:
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/useradd", "-m", UBUNTU_USER], check=True, timeout=30
+        )
+        with (UBUNTU_HOME / ".profile").open("a") as profile_file:
+            profile_file.write("PATH=$PATH:/home/ubuntu/.local/bin\n")
+        with (UBUNTU_HOME / ".bashrc").open("a") as bashrc_file:
+            bashrc_file.write("PATH=$PATH:/home/ubuntu/.local/bin\n")
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/groupadd", MICROK8S_GROUP], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", DOCKER_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", MICROK8S_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/sbin/usermod", "-aG", LXD_GROUP, UBUNTU_USER], check=True, timeout=30
+        )
+        subprocess.run(  # nosec: B603
+            ["/usr/bin/chmod", "777", "/usr/local/bin"], check=True, timeout=30
+        )
+    except subprocess.SubprocessError as exc:
+        raise SystemUserConfigurationError from exc
+
+
+YQ_DOWNLOAD_URL_TMPL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/yq_linux_{BIN_ARCH}"
+)
+YQ_BINARY_CHECKSUM_URL = "https://github.com/mikefarah/yq/releases/latest/download/checksums"
+YQ_CHECKSUM_HASHES_ORDER_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/checksums_hashes_order"
+)
+YQ_EXTRACT_CHECKSUM_SCRIPT_URL = (
+    "https://github.com/mikefarah/yq/releases/latest/download/extract-checksum.sh"
+)
+
+
+class ExternalPackageInstallError(Exception):
+    """Represents an error installilng external packages."""
+
+
+def _validate_checksum(file: Path, expected_checksum: str) -> bool:
+    """Validate the checksum of a given file.
+
+    Args:
+        file: The file to calculate checksum for.
+        expected_checksum: The expected file checksum.
+
+    Returns:
+        True if the checksums match. False otherwise.
+    """
+    sha256 = hashlib.sha256()
+    sha256.update(file.read_bytes())
+    return sha256.hexdigest() == expected_checksum
+
+
+BIN_ARCH_MAP: dict[Arch, str] = {Arch.ARM64: "arm64", Arch.X64: "amd64"}
+
+
+def _install_external_packages(arch: Arch) -> None:
+    """Install packages outside of apt.
+
+    Installs yarn, yq.
+
+    Args:
+        arch: The architecture to download binaries for. #TODO check bin arch
+
+    Raises:
+        ExternalPackageInstallError: If there was an error installing external package.
+    """
+    try:
+        subprocess.run(
+            ["/usr/bin/npm", "install", "--global", "yarn"], check=True, timeout=60 * 5
```

Comment:
There's a potential security risk here, npm is subject to toolchain attacks

Reply:
We also don't quite have a yarn snap: https://forum.snapcraft.io/t/classic-confinement-for-yarn/2686
Should this be built from source within this app as well?

Reply:
Don't know if that'll be trivial, but that's a risk we should be aware of